### PR TITLE
add ability to write more options for a claim

### PR DESCRIPTION
**Pull request by Flip-Liquid**
Addresses #52

### CODE REVIEW

File: src/OptionSettlement.sol [hunk 1]
```diff
@@ -188,31 +188,52 @@ contract OptionSettlementEngine is ERC1155, IOptionSettlementEngine {
 
     /// @inheritdoc IOptionSettlementEngine
     function write(uint256 optionId, uint112 amount) external returns (uint256 claimId) {
-        if (amount == 0) {
-            revert AmountWrittenCannotBeZero();
-        }
+        /// claimId == 0 signifies that a new claim NFT should be minted
```
Comment (0xAlcibiades) [inline src/OptionSettlement.sol:191]: Where is this claimid coming from? Will it ever be 0?

File: src/OptionSettlement.sol [hunk 2]
```diff
@@ -188,31 +188,52 @@ contract OptionSettlementEngine is ERC1155, IOptionSettlementEngine {
...
+        return write(0, optionId, amount);
+    }
 
+    /// @inheritdoc IOptionSettlementEngine
+    function write(uint256 claimId, uint256 optionId, uint112 amount) public returns (uint256) {
         (uint160 _optionId, uint96 claimIndex) = getDecodedIdComponents(optionId);
+
+        // claim index must be zero in lower 96b for provided option Id
```
Comment (0xAlcibiades) [inline src/OptionSettlement.sol:199]: Why? I don't think this is true.

File: src/OptionSettlement.sol [hunk 3]
```diff
@@ -188,31 +188,52 @@ contract OptionSettlementEngine is ERC1155, IOptionSettlementEngine {
...
+            revert EncodedOptionIdInClaimIdDoesNotMatchProvidedOptionId(claimId, optionId);
         }
 
-        uint256 rxAmount = amount * optionRecord.underlyingAmount;
-        uint256 fee = ((rxAmount / 10000) * feeBps);
-        address underlyingAsset = optionRecord.underlyingAsset;
+        Option storage optionRecord = _writeOptions(_optionId, amount);
+        uint256 mintClaimNft = 0;
+
+        if (claimId == 0) {
```
Comment (0xAlcibiades) [inline src/OptionSettlement.sol:212]: Again here, the assumtion about the zeroness of claimID, we want to support many claims per option type.

File: src/OptionSettlement.sol (hunk 1 above)
Comment (0xAlcibiades) [reply src/OptionSettlement.sol:191]: A new claim should always be minted if one is not provided.

File: src/OptionSettlement.sol [hunk 4]
```diff
@@ -188,31 +188,52 @@ contract OptionSettlementEngine is ERC1155, IOptionSettlementEngine {
...
+            // create new claim
```
Comment (0xAlcibiades) [inline src/OptionSettlement.sol:213]: In the case where a claim is provided, it must exist, elsewise the user must use the non overloaded version of this function.

File: src/OptionSettlement.sol [hunk 5]
```diff
@@ -188,31 +188,52 @@ contract OptionSettlementEngine is ERC1155, IOptionSettlementEngine {
...
     function write(uint256 optionId, uint112 amount) external returns (uint256 claimId) {
-        if (amount == 0) {
-            revert AmountWrittenCannotBeZero();
-        }
+        /// claimId == 0 signifies that a new claim NFT should be minted
+        return write(0, optionId, amount);
+    }
 
+    /// @inheritdoc IOptionSettlementEngine
+    function write(uint256 claimId, uint256 optionId, uint112 amount) public returns (uint256) {
```
Comment (0xAlcibiades) [inline src/OptionSettlement.sol:196]: claimId, being an overload, should be the 3rd argument here, for clarity.

**Review by 0xAlcibiades — CHANGES_REQUESTED** [review]

File: src/OptionSettlement.sol (hunk 1 above)
Comment (Flip-Liquid) [reply src/OptionSettlement.sol:191]: this is what's happening here, I think the comment is just misleading

File: src/OptionSettlement.sol (hunk 2 above)
Comment (Flip-Liquid) [reply src/OptionSettlement.sol:199]: the 'why' is bc it wouldn't make sense to call write(<claim id>, <amount>) the first argument of write() should always be an option ID. option IDs, from our encoding scheme, necessarily have the lower 96b set to zero. The claimIndex referred to here is just those lower 96b.

Comment (Flip-Liquid) [reply src/OptionSettlement.sol:199]: an improvement here would be to retitle the claimIndex var to make this clearer

File: src/OptionSettlement.sol (hunk 3 above)
Comment (Flip-Liquid) [reply src/OptionSettlement.sol:212]: I think the variable naming is what's confusing here. The claim ID is not assumed to be zero. Rather, we require that the lower 96b of the provided option ID are zero (perhaps mislabeled as `claimIndex`, as to be distinguisehd from the provided `claimId`). Here, if the claimID is zero, we create a new claim.

File: src/OptionSettlement.sol (hunk 4 above)
Comment (Flip-Liquid) [reply src/OptionSettlement.sol:213]: that's not the way it's written currently, so just want to describe what's there currently and see if you disagree with the design.

write(option, amount) writes `amount` of `option`, and puts this lot into a new `Claim`.

write(claimId, option, amount) writes `amount` of `option`, and if claimID is not zero, adds this amount to the existing claimID. If claimId is zero, then this is taken to mean "a claim does not exist, go create one".

write(option, amount) is just a call to [overloaded] write(0, option, amount)

**Review by 0xAlcibiades — APPROVED** [review]

**Pull request merged**